### PR TITLE
Additional fields

### PR DESCRIPTION
**Pull request by gcsantos-gpa**

### CODE REVIEW

**Comment by gcsantos-gpa** [comment]
I'll rebase after approval

File: Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/AdditionalFields/ByAdditionalField.tsx [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+    };
+    const [record, setRecord] = React.useState<SystemCenter.Types.AdditionalField>(emptyRecord);
+
+    const AdditionalFieldSearchField: Array<Search.IField<SystemCenter.Types.AdditionalField>> = [
+        { label: 'Name', key: 'FieldName', type: 'string', isPivotField: false },
+        {
+            label: 'Parent Type', key: 'ParentTable', type: 'enum', isPivotField: false,
+            enum: [
+                { Value: 'Meter', Label: 'Meter' },
+                { Value: 'Location', Label: 'Location' },
```
Comment (elwills) [inline Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/AdditionalFields/ByAdditionalField.tsx:71]:
```suggestion
                { Value: 'Location', Label: 'Substation' },
```

File: Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/AdditionalFields/ByAdditionalField.tsx [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+                { Value: 'Customer', Label: 'Customer' },
+                { Value: 'Company', Label: 'Company' },
+                { Value: 'Asset', Label: 'Asset' },
+                { Value: 'Line', Label: 'Line' },
+                { Value: 'LineSegment', Label: 'LineSegment' },
```
Comment (elwills) [inline Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/AdditionalFields/ByAdditionalField.tsx:76]:
```suggestion
                { Value: 'LineSegment', Label: 'Line Segment' },
```

File: Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/AdditionalFields/ByAdditionalField.tsx [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+                { Value: 'Breaker', Label: 'Breaker' },
+                { Value: 'CapacitorBank', Label: 'CapacitorBank' },
```
Comment (elwills) [inline Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/AdditionalFields/ByAdditionalField.tsx:78]:
```suggestion
                { Value: 'CapacitorBank', Label: 'Capacitor Bank' },
```

File: Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/AdditionalFields/ByAdditionalField.tsx [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+                { Value: 'Transformer', Label: 'Transformer' },
+                { Value: 'CapacitorBankRelay', Label: 'CapacitorBankRelay' },
```
Comment (elwills) [inline Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/AdditionalFields/ByAdditionalField.tsx:80]:
```suggestion
                { Value: 'CapacitorBankRelay', Label: 'Capacitor Bank Relay' },
```

File: Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/CommonComponents/FilterSelect.tsx [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+                Type={(props.Single ?? false) ? 'single' : 'multiple'}
+                StorageID={props.StorageID}
+                Columns={[
+                    { key: 'Name', field: 'Name', label: 'Name', headerStyle: { width: 'auto' }, rowStyle: { width: 'auto' } },
+                    { key: 'LocationKey', field: 'LocationKey', label: 'Key', headerStyle: { width: '30%' }, rowStyle: { width: '30%' } },
+                    { key: 'Meters', field: 'Meters', label: 'Meters', headerStyle: { width: '10%' }, rowStyle: { width: '10%' } },
+                    { key: 'Assets', field: 'Assets', label: 'Assets', headerStyle: { width: '10%' }, rowStyle: { width: '10%' } },
+                    { key: 'Scroll', label: '', headerStyle: { width: 17, padding: 0 }, rowStyle: { width: 0, padding: 0 } },
+                ]}
+                Title={props.Title === undefined ? "Select Locations" : props.Title}
```
Comment (elwills) [inline Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/CommonComponents/FilterSelect.tsx:176]:
```suggestion
                Title={props.Title === undefined ? "Select Substations" : props.Title}
```

File: Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/ExternalDB/ExternalDB.tsx [hunk 6]
```diff
@@ -22,85 +22,154 @@
...
+            type: "POST",
+            url: `${homePath}api/SystemCenter/ExternalDatabases/UnscheduledUpdate`,
+            contentType: "application/json; charset=utf-8",
+            data: JSON.stringify(record),
+            dataType: 'json',
+            cache: false,
+            async: true
+        });
+        handle.done(() => {
+            setPopupMessage("Unscheduled Update has Ran!");
```
Comment (elwills) [inline Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/ExternalDB/ExternalDB.tsx:117]:
```suggestion
            setPopupMessage("Unscheduled update successful.");
```

File: Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/ExternalDB/ExternalDBTableFieldForm.tsx [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+            return props.Record.FieldName != null && props.Record.FieldName.length > 0 && props.Record.FieldName.length <= 200;;
+
+        return true;
+    }
+
+    return (
+        <form>
+            <Input<SystemCenter.Types.AdditionalField> Record={props.Record} Field={'FieldName'} Label={'Name'} Feedback={'A Name of less than 200 characters is required.'} Valid={Valid} Setter={props.Setter} />
+            <Select<SystemCenter.Types.AdditionalField> Record={props.Record} Field={'ParentTable'} Label={'Parent Type'} Setter={props.Setter}
+                Options={[{ Value: 'Meter', Label: 'Meter' }, { Value: 'Location', Label: 'Location' }, { Value: 'Customer', Label: 'Customer' }, { Value: 'Company', Label: 'Company' }, { Value: 'Asset', Label: 'Asset' },
```
Comment (elwills) [inline Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/ExternalDB/ExternalDBTableFieldForm.tsx:59]:
```suggestion
                Options={[{ Value: 'Meter', Label: 'Meter' }, { Value: 'Location', Label: 'Substation' }, { Value: 'Customer', Label: 'Customer' }, { Value: 'Company', Label: 'Company' }, { Value: 'Asset', Label: 'Asset' },
```

File: Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/ExternalDB/ExternalDBTableFields.tsx [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+                        selected.filter((s) => origFields.findIndex((o) => o.ID === s.ID) < 0).forEach((f) => AssociateField(f));
+                        origFields.filter((o) => selected.findIndex((s) => s.ID === o.ID) < 0).forEach((f) => DisassociateField(f));
+                    }
+                }}
+                Show={showExisting}
+                Type={'multiple'}
+                Columns={[
+                    { key: 'FieldName', field: 'FieldName', label: 'Name', headerStyle: { width: 'auto' }, rowStyle: { width: 'auto' } },
+                    { key: 'ParentTable', field: 'ParentTable', label: 'Parent Type', headerStyle: { width: 'auto' }, rowStyle: { width: 'auto' } },
+                    { key: 'Type', field: 'Type', label: 'Type', headerStyle: { width: 'auto' }, rowStyle: { width: 'auto' } },
```
Comment (elwills) [inline Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/ExternalDB/ExternalDBTableFields.tsx:240]: Should we make this label `Field Type` to better differentiate from `Parent Type`?

Comment (elwills) [reply Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/ExternalDB/ExternalDBTableFields.tsx:240]: Or `Value Type` or something else. I'm not picky, just seems like something that could be confusing to some.

Comment (pdinsdale) [reply Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/ExternalDB/ExternalDBTableFields.tsx:240]: Yeah `Field Type` sounds good

File: Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/AdditionalFields/ByAdditionalField.tsx [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+                { Value: 'CapacitorBank', Label: 'Capacitor Bank' },
+                { Value: 'Transformer', Label: 'Transformer' },
+                { Value: 'CapacitorBankRelay', Label: 'Capacitor Bank Relay' },
+                { Value: 'DER', Label: 'DER' }
+            ]
+        },
+        // Todo: maybe override query results to allow for ExternalDBTableID to be used as a boolean?
+        // { label: 'External DB', key: 'ExternalDBTableID', type: 'boolean', isPivotField: false },
+        {
+            label: 'Type', key: 'Type', isPivotField: false, type: 'enum',
```
Comment (elwills) [inline Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/AdditionalFields/ByAdditionalField.tsx:87]:
```suggestion
            label: 'Field Type', key: 'Type', isPivotField: false, type: 'enum',
```

**Review by elwills — APPROVED** [review]

**Pull request merged**